src/shared/tpm2-util.c

```diff
@@ -0,0 +1,991 @@
+                                DLSYM_ARG(Esys_PolicyGetDigest),
+                                DLSYM_ARG(Esys_PolicyPCR),
+                                DLSYM_ARG(Esys_StartAuthSession),
+                                DLSYM_ARG(Esys_Startup),
+                                DLSYM_ARG(Esys_Unseal),
+                                NULL);
+                if (r < 0)
+                        return r;
+
+                libtss2_esys_dl = TAKE_PTR(dl);
+                k++;
+        }
+
+        if (!libtss2_rc_dl) {
+                _cleanup_(dlclosep) void *dl = NULL;
+
+                dl = dlopen("libtss2-rc.so.0", RTLD_LAZY);
+                if (!dl)
+                        return log_debug_errno(SYNTHETIC_ERRNO(EOPNOTSUPP),
+                                               "TPM2 support is not installed: %s", dlerror());
+
+                r = dlsym_many_and_warn(
+                                dl,
+                                LOG_DEBUG,
+                                DLSYM_ARG(Tss2_RC_Decode),
+                                NULL);
+                if (r < 0)
+                        return r;
+
+                libtss2_rc_dl = TAKE_PTR(dl);
+                k++;
+        }
+
+        if (!libtss2_mu_dl) {
+                _cleanup_(dlclosep) void *dl = NULL;
+
+                dl = dlopen("libtss2-mu.so.0", RTLD_LAZY);
+                if (!dl)
+                        return log_debug_errno(SYNTHETIC_ERRNO(EOPNOTSUPP),
+                                               "TPM2 support is not installed: %s", dlerror());
+
+                r = dlsym_many_and_warn(
+                                dl,
+                                LOG_DEBUG,
+                                DLSYM_ARG(Tss2_MU_TPM2B_PRIVATE_Marshal),
+                                DLSYM_ARG(Tss2_MU_TPM2B_PRIVATE_Unmarshal),
+                                DLSYM_ARG(Tss2_MU_TPM2B_PUBLIC_Marshal),
+                                DLSYM_ARG(Tss2_MU_TPM2B_PUBLIC_Unmarshal),
+                                NULL);
+                if (r < 0)
+                        return r;
+
+                libtss2_mu_dl = TAKE_PTR(dl);
+                k++;
+        }
+
+        return k;
+}
+
+struct tpm2_context {
+        ESYS_CONTEXT *esys_context;
+        void *tcti_dl;
+        TSS2_TCTI_CONTEXT *tcti_context;
+};
+
+static void tpm2_context_destroy(struct tpm2_context *c) {
+        assert(c);
+
+        if (c->esys_context)
+                sym_Esys_Finalize(&c->esys_context);
+
+        c->tcti_context = mfree(c->tcti_context);
+
+        if (c->tcti_dl) {
+                dlclose(c->tcti_dl);
+                c->tcti_dl = NULL;
+        }
+}
+
+static inline void Esys_Finalize_wrapper(ESYS_CONTEXT **c) {
+        /* A wrapper around Esys_Finalize() for use with _cleanup_(). Only reasons we need this wrapper is
+         * because the function itself warn logs if we'd pass a pointer to NULL, and we don't want that. */
```

Comment:
Very annoying when libraries do this. Maybe worth filing an issue upstream?

Reply:
Well, I find that annoying, but there are two schools of thought about that, and it feels like getting into a political fight, requesting them to follow our style that freeing/destroying/finalizing NULL stuff should be a NOP...

I mean we can try...

Reply:
I have posted this now: https://github.com/tpm2-software/tpm2-tss/issues/1942

Let's see if that goes anywhere.